analyzers/src/SonarAnalyzer.Common/Rules/CheckFileLicenseBase.cs

```diff
@@ -18,31 +18,61 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
 
 namespace SonarAnalyzer.Rules
 {
     public abstract class CheckFileLicenseBase : ParameterLoadingDiagnosticAnalyzer
     {
         internal const string DiagnosticId = "S1451";
-        protected const string MessageFormat = "Add or update the header of this file.";
-
         internal const string HeaderFormatRuleParameterKey = "headerFormat";
         internal const string HeaderFormatPropertyKey = nameof(HeaderFormat);
-        public abstract string HeaderFormat { get; set; }
-
         internal const string IsRegularExpressionRuleParameterKey = "isRegularExpression";
         internal const string IsRegularExpressionPropertyKey = nameof(IsRegularExpression);
         internal const string IsRegularExpressionDefaultValue = "false";
-        [RuleParameter(IsRegularExpressionRuleParameterKey, PropertyType.Boolean,
-            "Whether the headerFormat is a regular expression.", IsRegularExpressionDefaultValue)]
+        protected const string MessageFormat = "Add or update the header of this file.";
+
+        private readonly DiagnosticDescriptor rule;
+
+        protected abstract ILanguageFacade Language { get; }
+        public abstract string HeaderFormat { get; set; }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        [RuleParameter(IsRegularExpressionRuleParameterKey, PropertyType.Boolean, "Whether the headerFormat is a regular expression.", IsRegularExpressionDefaultValue)]
         public bool IsRegularExpression { get; set; } = bool.Parse(IsRegularExpressionDefaultValue);
 
+        protected CheckFileLicenseBase() =>
+            rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, Language.RspecResources);
+
+        protected override void Initialize(ParameterLoadingAnalysisContext context) =>
+            context.RegisterSyntaxTreeActionInNonGenerated(Language.GeneratedCodeRecognizer, c =>
+                {
+                    if (HeaderFormat == null)
+                    {
+                        return;
+                    }
+
+                    if (IsRegularExpression && !IsRegexPatternValid(HeaderFormat))
+                    {
+                        throw new InvalidOperationException($"Invalid regular expression: {HeaderFormat}");
+                    }
+
+                    var firstNode = c.Tree.GetRoot().ChildTokens().FirstOrDefault().Parent;
+                    if (!HasValidLicenseHeader(firstNode))
+                    {
+                        var properties = CreateDiagnosticProperties();
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, Location.Create(c.Tree, TextSpan.FromBounds(0, 0)), properties));
+                    }
+                });
+
         protected static bool IsRegexPatternValid(string pattern)
```

Comment:
can be private

Reply:
Nice, almost everything is private now :)